src/hardware/serialport/serialmouse.cpp

```diff
@@ -0,0 +1,453 @@
+        return;
+    }
+
+    std::string rate_string;
+    use_default = !cmd->FindStringBegin("rate:", rate_string, false);
+
+    if (rate_string == "smooth" || use_default)
+        // This is roughly equivalent of PS/2 mouse with sampling rate between 120 and 250 Hz,
+        // depends on mouse protocol, concrete events (reporting middle button state change or
+        // wheel movement needs more bandwidth for Logitech and wheel mice), and sometimes even
+        // driver sophistication (for Mouse Systems mouse a badly written driver can waste half
+        // of the movement sampling rate)
+        smooth_div = 5;
+    else if (rate_string != "normal") {
+        LOG_ERR("MOUSE (COM%d): Invalid rate '%s'", port_num, rate_string.c_str());
+        return;
+    }
+
+    setType(config_type);
+
+    CSerial::Init_Registers();
+    setRI(false);
+    setDSR(false);
+    setCD(false);
+    setCTS(false);
+
+    InstallationSuccessful = true;
+
+    MouseSER_RegisterListener(this);
+}
+
+CSerialMouse::~CSerialMouse() {
+    MouseSER_UnRegisterListener(this);
+    removeEvent(SERIAL_TX_EVENT); // clear events
+    setType(MouseType::NO_MOUSE);
+}
+
+void CSerialMouse::setType(MouseType type) {
+    if (type != mouse_type) {
+        mouse_type = type;
+        const char* type_name = nullptr;
+        switch (type) {
+         case MouseType::NO_MOUSE: // just to print out log in the destructor
+            type_name            = "(none)";
+            break;
+        case MouseType::MICROSOFT:
+            type_name            = "Microsoft, 2 buttons";
+            mouse_bytelen        = 7;
+            mouse_has_3rd_button = false;
+            mouse_has_wheel      = false;
+            break;
+        case MouseType::LOGITECH:
+            type_name            = "Logitech, 3 buttons";
+            mouse_bytelen        = 7;
+            mouse_has_3rd_button = true;
+            mouse_has_wheel      = false;
+            break;
+        case MouseType::WHEEL:
+            type_name            = "wheel, 3 buttons";
+            mouse_bytelen        = 7;
+            mouse_has_3rd_button = true;
+            mouse_has_wheel      = true;
+            break;
+        case MouseType::MOUSE_SYSTEMS:
+            type_name            = "Mouse Systems, 3 buttons";
+            mouse_bytelen        = 8;
+            mouse_has_3rd_button = true;
+            mouse_has_wheel      = false;
+            break;      
+        default:
+            unimplemented();
+            break;
+        }
+
+        if (type_name)
+            LOG_MSG("MOUSE (COM%d): %s", port_num, type_name);
+    }
+}
+
+void CSerialMouse::abortPacket() {
+    packet_len          = 0;
+    xmit_idx            = 0xff;
+    xmit_2part          = false;
+    xmit_another_move   = false;
+    xmit_another_button = false;
+}
+
+void CSerialMouse::clearCounters() {
+    mouse_delta_x       = 0;
+    mouse_delta_y       = 0;
+    mouse_delta_w       = 0;  
```

Comment:
minor: formatting

Reply:
Done, but the script will be used on the file nevertheless.